File: crates/nil-core-macros/src/building/storage.rs
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

use proc_macro::TokenStream;
use quote::quote;
use syn::DeriveInput;

pub fn impl_storage(ast: &DeriveInput) -> TokenStream {
  let name = &ast.ident;
  let stream = quote! {
    mod __impl_storage {
      use super::#name;
      use crate::error::Result;
      use crate::infrastructure::storage::{
        Storage,
        StorageCapacity,
        StorageId,
        StorageStatsTable
      };

      impl Storage for #name {
        fn storage_id(&self) -> StorageId {
          Self::STORAGE_ID
        }

        fn capacity(&self, stats: &StorageStatsTable) -> Result<StorageCapacity> {
          Ok(stats.get(self.level)?.capacity)
        }

        fn min_capacity(&self) -> StorageCapacity {
          Self::MIN_CAPACITY
        }

        fn max_capacity(&self) -> StorageCapacity {
          Self::MAX_CAPACITY
        }
      }
    }
  };

  stream.into()
}

File: app/src/lib/storage.ts
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

import { getCurrentScope } from "vue";
import { useRoute } from "vue-router";
import { runWithContext } from "@tb-dev/vue";

export function prependKey(key: string): string {
  if (getCurrentScope()) {
    return runWithContext(() => {
      const route = useRoute();
      if (route.name && typeof route.name === "string") {
        return `nil:${route.name}:${key}`;
      }
      else {
        return `nil:${key}`;
      }
    });
  }
  else {
    return key;
  }
}
